UnitTest/DataStructuresTests/BTreeTest.cs

```diff
@@ -0,0 +1,211 @@
+            bTree.Insert(15);
+            bTree.Insert(25);
+            bTree.Insert(39);
+
+            // The tree now looks like this:
+            /***************************************
+             **
+             **                      [20 , 40]
+             **                     /    |    \
+             **                    /     |     \
+             **         [5, 10, 15]      |      [50 , 60 , 70]
+             **                          |
+             **                   [25, 30, 35, 39]
+             **
+             ***************************************
+             */
+
+            Assert.Equal(2, bTree.Search(20).Keys.Count);
+            Assert.Equal(2, bTree.Search(40).Keys.Count);
+            Assert.Null(bTree.Search(41));
+            Assert.Equal(3, bTree.Search(5).Keys.Count);
+            Assert.Equal(4, bTree.Search(25).Keys.Count);
+        }
+
+        [Fact]
+        public static void DoDeleteTest()
+        {
+            // Build a base tree
+            BTree<int> bTree = new BTree<int>(4);
+            bTree.Insert(10);
+            bTree.Insert(30);
+            bTree.Insert(20);
+            bTree.Insert(50);
+            bTree.Insert(40);
+            bTree.Insert(60);
+            bTree.Insert(70);
+            bTree.Insert(35);
+            bTree.Insert(5);
+            bTree.Insert(15);
+            bTree.Insert(25);
+            bTree.Insert(39);
+
+            // The tree now looks like this:
+            /***************************************
+             **
+             **                      [20 , 40]
+             **                     /    |    \
+             **                    /     |     \
+             **         [5, 10, 15]      |      [50 , 60 , 70]
+             **                          |
+             **                   [25, 30, 35, 39]
+             **
+             ***************************************
+             */
+            // First. assert the shape.
+            Assert.Equal(2, bTree.Search(20).Keys.Count);
+            Assert.Equal(2, bTree.Search(40).Keys.Count);
+            Assert.Null(bTree.Search(41));
+            Assert.Equal(3, bTree.Search(5).Keys.Count);
+            Assert.Equal(4, bTree.Search(25).Keys.Count);
+
+            // Now, remove a key from the left-most child.
+            bTree.Remove(5);
+
+            // The tree now looks like this:
+            /***************************************
+             **
+             **                              [40]
+             **                             /    \
+             **                            /      \
+             ** [10, 15, 20, 25, 30, 35, 39]       [50 , 60 , 70]
+             ** 
+             ***************************************
+             */
+
+            // The tree should now be rooted around 40, with the left child full.
+            Assert.Equal(null, bTree.Search(5));
+            Assert.Equal(2, bTree.Root.Children.Count);
```

Comment:
Is this assertion true? It looks, from your diagram, that the right node has 3 children and not 2.

Reply:
Nevermind. Misread the tests.